pkg/ruler/ruler.go

```diff
@@ -409,12 +394,46 @@ func (r *Ruler) loadRules(ctx context.Context) {
 		if _, exists := configs[user]; !exists {
 			go mngr.Stop()
 			delete(r.userManagers, user)
-			level.Info(util.Logger).Log("msg", "deleting rule manager", "user", user)
+			level.Info(r.logger).Log("msg", "deleting rule manager", "user", user)
 		}
 	}
 
 }
 
+// syncManager maps the rule files to disk, detects any changes and will create/update the
+// the users Prometheus Rules Manager
+func (r *Ruler) syncManager(ctx native_ctx.Context, user string, groups store.RuleGroupList) {
+	r.userManagerMtx.Lock()
```

Comment:
Before we were locking only around `r.userManagers` access, while now we lock around the entire function. Could you explain why? If you confirm it's the indented behaviour, adding a comment before this `Lock()` explaining the reason may be useful for future readers (it's unclear to me).

Reply:
As soon as `update, files, err := r.mapper.MapRules(user, groups.Formatted())` is called we need to lock until we call `manager.Update` for each user. Otherwise it there is a race condition that can make itself know in tests.

Assume there are two calls to `syncManager`:
  - `CallOne` maps files and returns true to update user managers
  - `CallTwo` maps files and returns false to update user managers
  - `CallTwo` returns and wherever it is called assumes that the running managers for each user are up to date, but they aren't because `CallOne` hasn't completed yet.

It is an issue that will only pop up in tests, since we need to force a sync instead of letting it occur in a background task. The other option would be adding a sleep to tests. However, I generally like to avoid doing that if need be. I also think it is very likely syncManagers will be called concurrently at some point. Specifically, as part of the Ruler lifecycle, we would want a new ruler joining/leaving the ring to trigger a sync in all of the other rulers.

I'll add a comment to make what is going on clear. I just wanted to write a more detailed explanation here.